### PR TITLE
use Array for regions instead of embedded linked list

### PR DESCRIPTION
**Pull request by WalterBright**
Using an array makes traversal more cache friendly, and is simpler code.

A future PR could then sort it, to make `container()` use a binary rather than linear search.

### CODE REVIEW

**Comment by WalterBright** [comment]
@rainers your advice please

File: src/dmd/root/region.d [hunk 1]
```diff
@@ -69,16 +75,18 @@ struct Region
...
         {
-            next = *cast(void**)h;
-            memset(h, 0xFC, ChunkSize);
-            .free(h);
+            /* Free the memory
+             */
+            foreach (h; array)
+            {
+                memset(h, 0xFC, ChunkSize); // 0xFC is larger than TOK.max
+                .free(h);
```
Comment (rainers) [inline src/dmd/root/region.d]: This `free` is wrong here. Memory should only be stomped, not freed.

File: src/dmd/root/region.d [hunk 2]
```diff
@@ -18,21 +18,25 @@ import core.stdc.string;
...
+  private:
+
+    Array!(void*) array; // array of chunks
+    int used;            // number of chunks used in array[]
+    void[] available;    // slice of chunk that's available to allocate
 
-    enum ChunkSize = 4096;
-    enum OverheadSize = 16;
+    enum ChunkSize = 4096 * 1024;
+    enum OverheadSize = 0;
```
Comment (rainers) [inline src/dmd/root/region.d]: I'd recommend removing `OverheadSize` from this version, it just creates confusion.

File: src/dmd/root/region.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/root/region.d]: It's not compiled in at the moment.

File: src/dmd/root/region.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/root/region.d]: ok

File: src/dmd/root/region.d (hunk 1 above)
Comment (rainers) [reply src/dmd/root/region.d]: I know but it will bite you if the code is enabled for debugging. There is no reason to keep it.

Comment (WalterBright) [reply src/dmd/root/region.d]: ok

**Review by rainers — APPROVED** [review]

**Pull request merged**